Use Glob in PUT file paths and Allow for Multi File Upload

colin99d opened the pull request:
This is a work in progress, I would be very thankful for any feedback.

By adding the GLOB package, file paths like the one below are now allowed:
`api.exec(&format!("PUT file://{} @%OSCAR_AGE_MALE;", "/Users/colin/test-snowflake/data/*.csv"))`

This PR uses Tokio tasks to manage spawn a thread for each individual file.

This PR also moves libraries only required for the PUT command into the `file` feature

colin99d committed:
Refactor PUT to avoid sending files before validating all filenames are valid

colin99d committed:
Merge branch 'main' into MultiFile

colin99d committed:
Added glob with file paths

colin99d committed:
Works, but really dirty and no speed optimization

colin99d committed:
Cleaned up multi file upload

colin99d committed:
Made file an optional feature

colin99d committed:
Cleaned up for an MVP

colin99d committed:
FMT

colin99d committed:
Added fix

colin99d committed:
Limit concurrency

andrusha commented on snowflake-api/src/lib.rs:
```diff
@@ -225,22 +245,53 @@ impl SnowflakeApi {
             .with_token(info.creds.aws_token)
             .build()?;
 
+        let s3_mutex = Arc::new(Mutex::new(s3));
```
Wouldn't using mutex here defeat the purpose of uploading files in parallel? Since you're going to lock IO intensive task into a single thread

colin99d replied:
I think your right! Do you have any suggestions for a better solution? This is probably why I didn't see performance gains.

andrusha replied:
It appears to me to be thread safe, I guess you will be ok without Mutex

colin99d replied:
This sped up the code 5 times!

andrusha commented on snowflake-api/src/lib.rs:
```diff
@@ -225,22 +245,53 @@ impl SnowflakeApi {
-            let fs = LocalFileSystem::new().get(&src_path).await?;
+        let bucket_path = bucket_path.to_string();
+        let mut set = JoinSet::new();
+        for src_path in final_source_locs {
```
I think using Go and .NET official implementations as reference is a good idea here. 
https://github.com/snowflakedb/snowflake-connector-net/blob/7d478f0eaeb71493b878a47f87193df4e69c39b2/Snowflake.Data/Core/FileTransfer/SFFileTransferAgent.cs#L205-L217

The application should take into account if parallel uploading is supported for given target and obey threshold:
https://github.com/mycelial/snowflake-rs/blob/main/snowflake-api/src/responses.rs#L197-L201

colin99d replied:
So for right now I am using the Snowflake recommended limits of 64MB for file size and 4 threads for concurrency. I did not see a way to easily get the values you have linked here into the function, do you have any suggestions?

andrusha commented on snowflake-api/Cargo.toml:
```diff
@@ -32,14 +33,16 @@ reqwest-middleware = "0.2"
 uuid = { version = "1", features = ["v4"] }
+snowflake-jwt = { version = "0.3.0", optional = true }
+glob = { version = "0.3.1", optional = true }
+tokio = { version = "1.35", optional = true, features = ["macros", "rt-multi-thread"] }
```
It's bad practice to fix minor versions if you absolutely don't have to. Library users would be limited by this even if they have legitimate reasons to use specific version.

andrusha commented on snowflake-api/src/lib.rs:
```diff
@@ -225,22 +245,53 @@ impl SnowflakeApi {
+            let mutex1 = Arc::clone(&s3_mutex);
+            let bucket_path = bucket_path.clone();
+            while set.len() >= 5 {
+                set.join_next().await.unwrap().unwrap();
```
So only the last batch of 5 will get proper error handling and this crashes with exception? I think there could be nicer way to split tasks into batches of given size.

colin99d replied:
This doesnt seem to offer a performance improvement. I will just remove it for now.

colin99d committed:
Fixed glob

colin99d committed:
Removed unnecessary batching

colin99d committed:
Only upload small files sequentially

andrusha commented on snowflake-api/Cargo.toml:
```diff
@@ -13,6 +13,7 @@ version = "0.6.0"
 
 [features]
 cert-auth = ["dep:snowflake-jwt"]
+file = ["dep:glob", "dep:tokio", "dep:object_store"]
```
I think it's a bit self-defeating to hide all features by-default. The smallest library is the one which does nothing, but also the most useless one. Enabling reasonable amount of features and avoiding user surprise (getting an unsupported error in runtime) is a better way to go.

andrusha replied:
As for making tokio optional, `reqwest` needs it either way

andrusha commented on snowflake-api/Cargo.toml:
```diff
@@ -32,14 +33,16 @@ reqwest-middleware = "0.2"
 uuid = { version = "1", features = ["v4"] }
+snowflake-jwt = { version = "0.3.0", optional = true }
+glob = { version = "0.3", optional = true }
+tokio = { version = "1.35", optional = true, features = ["macros", "rt-multi-thread"] }
```
Why fix it to `1.35` specifically? Wouldn't it work on `1.36` or `1.25`?

colin99d replied:
Updated version to match the reqwest example.

andrusha commented on snowflake-api/src/helpers.rs:
```diff
@@ -0,0 +1,116 @@
+use crate::SnowflakeApiError;
```
This should probably be called `upload_files.rs`, helpers could mean literally anything and obscures the purpose

andrusha commented on snowflake-api/src/helpers.rs:
```diff
@@ -0,0 +1,116 @@
+        });
+    }
+    while let Some(res) = set.join_next().await {
+        let result: Result<(), SnowflakeApiError> = res.unwrap();
```
why unwrap?

andrusha commented on snowflake-api/src/helpers.rs:
```diff
@@ -0,0 +1,116 @@
+}
+
+/// This function uploads files in parallel, useful for files above the threshold
+pub async fn upload_files_sequential(
```
comment doesn't correspond to function name

andrusha commented on snowflake-api/src/helpers.rs:
```diff
@@ -0,0 +1,116 @@
+        let dest_path = format!("{}{}", bucket_path.clone(), filename.to_str().unwrap());
+        let dest_path = object_store::path::Path::parse(dest_path)?;
+        let src_path = object_store::path::Path::parse(src_path)?;
+        let fs = LocalFileSystem::new().get(&src_path).await?;
```
couldn't this be abstracted somehow?

andrusha commented on snowflake-api/src/lib.rs:
```diff
@@ -75,6 +80,9 @@ pub enum SnowflakeApiError {
     UnexpectedResponse,
+
+    #[error("Missing feature: {0}")]
+    MissingFeature(String),
```
failing runtime because of the compile flag is not a good idea

colin99d replied:
This is a tricky one because we can't wrap the whole function in a cfg because it depends the value of the regex. To keep things simple, should we just not have a separate file feature for now?

colin99d committed:
Updated

colin99d committed:
Merged

colin99d committed:
Various improvements

colin99d committed:
Continued clean up

colin99d committed:
Cleaned the code and fixed a bug

colin99d commented:
FYI: Until the limit to concurrency is implemented this PR is not ready for review. I am more than happy for your comments. But please know some of this code is stuff I plan to clean on my own (right now this is all being coded while I am on planes and busses).

colin99d committed:
Did a lot of cleaning

colin99d committed:
Abstracted file processing code

colin99d committed:
Merged

colin99d committed:
Cleaned up clippy

colin99d committed:
Limit concurrent threads

colin99d commented:
@andrusha besides the two questions I had this PR is good to go. I decided to create a way for the user to define the file size and concurrency thresholds. Let me know your thoughts!

andrusha committed:
Merge remote-tracking branch 'origin/main' into MultiFile

andrusha committed:
Use limit store for parallel uploads

Use spawn blocking for computationally intensive glob traverse
Use async tokio fs for metadata fetching
Move all PUT query logic into put.rs file
Delete files feature for now

andrusha committed:
Merge remote-tracking branch 'origin/main' into MultiFile

andrusha committed:
Use auth env for file transfer example

andrusha committed:
Add note on glob uploading support

dmzmk reviewed (approved):
lgtm

colin99d commented:
> lgtm

Спасибо!
